subsys/bluetooth/host/scan.c

```diff
@@ -279,97 +310,189 @@ static int start_le_scan_legacy(uint8_t scan_type, uint16_t interval, uint16_t w
 		return err;
 	}
 
-	err = bt_le_scan_set_enable(BT_HCI_LE_SCAN_ENABLE);
+	err = bt_le_scan_set_enable_int(BT_HCI_LE_SCAN_ENABLE,
+					p_param->options & BT_LE_SCAN_OPT_FILTER_DUPLICATE);
 	if (err) {
 		return err;
 	}
 
-	atomic_set_bit_to(bt_dev.flags, BT_DEV_ACTIVE_SCAN, active_scan);
-
 	return 0;
 }
 
-static int start_host_initiated_scan(bool fast_scan)
+static int start_scan(struct bt_le_scan_param *p_scan_param)
 {
-	uint16_t interval, window;
-
-	if (fast_scan) {
-		interval = BT_GAP_SCAN_FAST_INTERVAL;
-		window = BT_GAP_SCAN_FAST_WINDOW;
-	} else {
-		interval = CONFIG_BT_BACKGROUND_SCAN_INTERVAL;
-		window = CONFIG_BT_BACKGROUND_SCAN_WINDOW;
+	if (IS_ENABLED(CONFIG_BT_EXT_ADV) && BT_DEV_FEAT_LE_EXT_ADV(bt_dev.le.features)) {
+		return start_le_scan_ext(p_scan_param);
 	}
 
-	if (IS_ENABLED(CONFIG_BT_EXT_ADV) &&
-	    BT_DEV_FEAT_LE_EXT_ADV(bt_dev.le.features)) {
-		struct bt_hci_ext_scan_phy scan_phy_params;
-
-		scan_phy_params.type = BT_HCI_LE_SCAN_PASSIVE;
-		scan_phy_params.interval = sys_cpu_to_le16(interval);
-		scan_phy_params.window = sys_cpu_to_le16(window);
+	return start_le_scan_legacy(p_scan_param);
+}
 
-		/* Scan on 1M + Coded if the controller supports it*/
-		if (BT_FEAT_LE_PHY_CODED(bt_dev.le.features)) {
-			return start_le_scan_ext(&scan_phy_params, &scan_phy_params, 0);
-		} else {
-			return start_le_scan_ext(&scan_phy_params, NULL, 0);
-		}
+static uint32_t m_scan_check_if_state_allowed(enum scan_enabled_reason enable_reason)
+{
+	/* check if state is already set */
+	if (atomic_test_bit(m_scanner_state.scan_flags, enable_reason)) {
+		return -EAGAIN;
```

Comment:
Should this be 
```suggestion
		return -EALREADY;
```

`EAGAIN` IIRC indicates that it can retry later, but that's not really the case if it just keeps trying to enable scan.

Reply:
I will also change the next return value in this function. Please check if that makes sense to you.